### PR TITLE
Documentation for z_obj_driftice.c

### PR DESCRIPTION
**Pull request by AdamKiddle**
I needed to document this actor in order to support some glitch hunting I was doing so I thought I'd submit it as a PR!

### CODE REVIEW

File: src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c [hunk 1]
```diff
@@ -76,200 +87,257 @@ static InitChainEntry sInitChain[] = {
...
 
-    arg0->unk_00.unk_02 += arg0->unk_00.unk_04;
-    Math_StepToF(&arg0->unk_00.unk_08, arg0->unk_00.unk_0C, arg0->unk_00.unk_10);
+    oscillationState->currentArg += oscillationState->argRateOfChange;
+    Math_StepToF(&oscillationState->currentMagnitude, oscillationState->maxMagnitude,
+                 oscillationState->magnitudeRateOfChange);
 
-    sp20 += (Math_SinS(arg0->unk_00.unk_02) * arg0->unk_00.unk_08);
-    return sp20;
+    yOffset += (Math_SinS(oscillationState->currentArg) * oscillationState->currentMagnitude);
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c:148]:
```suggestion
    yOffset += Math_SinS(oscillationState->currentArg) * oscillationState->currentMagnitude;
```

File: src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c [hunk 2]
```diff
@@ -76,200 +87,257 @@ static InitChainEntry sInitChain[] = {
...
+        Math_StepToF(&dynamics->impactFromPlayer, 1.0f, 0.02f);
     } else {
-        Math_StepToS(&arg0->unk_02, 0, 0x96);
-        phi_s0 = 0;
-        Math_StepToS(&arg0->unk_06, 20, 7);
-        Math_StepToF(&arg0->unk_08, 0.0f, 0.02f);
+        Math_StepToS(&dynamics->precessionAngleRateOfChange, 0, 150);
+        targetTiltMagnitude2 = 0; // Tilt magnitude will decrease to zero over time
+        Math_StepToS(&dynamics->tiltMagnitude2RateOfChange, 20,
+                     7); // Tilt magnitude rate of change will quickly reach 20
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c:221]:
```suggestion
        // Tilt magnitude will decrease to zero over time
        targetTiltMagnitude2 = 0;
        
        // Tilt magnitude rate of change will quickly reach 20
        Math_StepToS(&dynamics->tiltMagnitude2RateOfChange, 20, 7); 
```
The comments going above will help the formatting here

File: src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c [hunk 3]
```diff
@@ -76,200 +87,257 @@ static InitChainEntry sInitChain[] = {
...
-    ObjDrifticeStruct3* temp_s0;
-    f32 temp_f20 = 0.0f;
+/*
+Updates the state of the dynamics associated with generic precession of platform (unrelated to player interaction).
+ Populates tiltMagnitude and precessionAngle
+
+Updates three oscillation states and sums their outputs into tiltMagnitude
+precessionAngle is the sum of two random angles which change over time. Always between -600 and 0.
+*/
+void ObjDriftice_GenericPrecessionPhysicsUpdate(ObjDrifticeGenericPrecessionDynamics* dynamics, ObjDriftice* this,
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c:261]: For both `ObjDriftice_GenericPrecessionPhysicsUpdate` and `ObjDriftice_PlayerRelatedPrecessionPhysicsUpdate`, could you add to the function comments what precession is just like what you have in `ObjDriftice_UpdateDynamics`:
`precession (rotational tilt/sway)`

File: src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c [hunk 4]
```diff
@@ -76,200 +87,257 @@ static InitChainEntry sInitChain[] = {
...
-    ObjDrifticeStruct3* temp_s0;
-    f32 temp_f22 = 0.0f;
-    s16 phi_s0;
+/*
+Updates dynamics of platform precession based on player interaction and populates tiltMagnitude and precessionAngle.
+    Value stored in tiltMagnitude approaches zero when player is not on platform.
+
+Magnitude is calculated based on two contributions.
+- tiltMagnitude1 = the combined output of two oscillation states * impactFromPlayer
+- tiltMagnitude2 is based on player position and contains a feedback loop where targetTiltMagnitude2 and
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c:192]:
```suggestion
- tiltMagnitude1 is the combined output of two oscillation states * impactFromPlayer
- tiltMagnitude2 is based on player position and contains a feedback loop where targetTiltMagnitude2 and
```
Consistency

File: src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c [hunk 5]
```diff
@@ -76,200 +87,257 @@ static InitChainEntry sInitChain[] = {
...
 
-void func_80A66C4C(ObjDrifticeStruct4* arg0, ObjDriftice* this, s16* arg2, s16* arg3) {
-    ObjDrifticeStruct3* temp_s0;
-    f32 temp_f20 = 0.0f;
+/*
+Updates the state of the dynamics associated with generic precession of platform (unrelated to player interaction).
+ Populates tiltMagnitude and precessionAngle
+
+Updates three oscillation states and sums their outputs into tiltMagnitude
+precessionAngle is the sum of two random angles which change over time. Always between -600 and 0.
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c:259]:
```suggestion
precessionAngle is the sum of two random angles which change over time. Always between -0x258 and 0.
```
We always try to have angles in hex

**Review by engineer124 — COMMENTED** [review]
Great work! Mostly just cleanup comments

**Comment by AdamKiddle** [comment]
Sorry, missed the emails for these updates. Thanks for the review, I'll get to making these changes soon!

**Comment by hirottau302** [comment]
Ok so I have the legened of Zelda majoras mask to the n64 and on day when I was trying to play my save or if try to make a new game the game would freeze on a black screen with a little red line at the top left of the screen and it kept on doing that and I do not know how to fix and I have tried everything I could think of in which is changing the battery and cleaning the components still it does th he same thing. Please if y'all know anything about the problem please let me know.

**Comment by altoidmerope294** [comment]
I'm almost to the snow temple but the part where you have to get past the gigantic Goron that blows snow and a ton of air at you to make you tumble off the cliff. How do I get past him to get to the snow temple?

**Comment by yemenalcyone230** [comment]
?yrotnevni ym ni eb ksam ytieD ecreiF eht dluow iiW ym no mor ot teser ot erew I fi os "dnE ehT" syas ti tub emag eht taeb I woN .ksam eht peek ot em wolla t'ndid hcihw emit fo gnos eht deyalp I ti tog yllanigiro I nehW ...tub eno siht rof diputs yllaer leef ot gniog ylbaborp m'I

**Comment by ginomeelectra202** [comment]
Getting Epona to
 get to the Great Bay Temple

**Comment by altoidmerope294** [comment]
I'm in Ikana Canyon and Tatl keeps saying something about a mask that can reveal the truth. The only masks I have are Bomb, Great Fairy, Scent, Goron, Don Gero, Bremen, Rabbit, Deku, Zora, and the Captain's Hat.The Lens of Truth seemed to have no effect.Is there a mask I missed or does one of the masks I have contain some additional ability?

**Comment by yemenalcyone230** [comment]
i can assume it has something to do with magic, I'm just curios though.

**Comment by hirottau302** [comment]
?tellaw reggib eht teg ro dnif I od erehw os fi dna seepur 002 neht erom yrrac uoy naC

**Comment by hirottau302** [comment]
.tih eno ni stop kaerb uoy sah aroZ a erehw eno ehT

**Comment by ginomeelectra202** [comment]
how did he leave if he fell through a hole? please no spoilers, just dumb it down so it has the least possible amount of spoilers.

**Comment by AngheloAlf** [comment]
Hello, this PR hasn't been updated in a while. Do you have plans to update this PR? If we haven't heard back by June 1st then we will take it over / close the PR.

File: src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c [hunk 6]
```diff
@@ -76,200 +87,257 @@ static InitChainEntry sInitChain[] = {
...
+    f32 zeroF = 0.0f;
+    f32 yOffset;
+
+    if (this->timerPlayerOnTop >= Y_OFFSET_CHANGE_TIME) {
+        yOffset = -8.0f;
+    } else if (this->timerPlayerOnTop <= -Y_OFFSET_CHANGE_TIME) {
+        yOffset = 0.0f;
+    } else if (this->timerPlayerOnTop > 0) {
+        yOffset = (Math_CosS(fabsf(this->timerPlayerOnTop) * (0x8000 * (1.0f / Y_OFFSET_CHANGE_TIME))) - 1.0f) *
+                  4.0f; // Decrease from 0 to -8
```
Comment (TrulyNaN) [inline src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c:118]: I think it's in the spirit of what engineer said, but try to put comments above code and not on the same line. Just wanted to show you an example where it occured.

File: src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c (hunk 2 above)
Comment (TrulyNaN) [reply src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c:221]: If someone knows what Math_StepToS does, I think the two "Tilt magnitude ..." comments can be removed. StepTo functions are quite common.

File: src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c [hunk 7]
```diff
@@ -325,129 +393,131 @@ void ObjDriftice_Destroy(Actor* thisx, PlayState* play) {
...
         this->dyna.actor.world.pos.x += this->dyna.actor.velocity.x;
         this->dyna.actor.world.pos.y += this->dyna.actor.velocity.y;
         this->dyna.actor.world.pos.z += this->dyna.actor.velocity.z;
     } else {
-        sp30 = true;
-        this->unk_164 += this->unk_168;
+        snapToPathPoint = true;
+        this->curPathPointIdx += this->pathDirection;
 
         if (1) {}
```
Comment (TrulyNaN) [inline src/overlays/actors/ovl_Obj_Driftice/z_obj_driftice.c:439]: If you already tried to remove this `if (1)` with the permuter, ignore this comment. Otherwise, can you see if it can be removed with the permuter? Don't waste too much time on this if it doesn't work quickly as it's not abnormal to have to resort to that for matching.

**Review by TrulyNaN — COMMENTED** [review]
Thanks for documenting the actor and good work! Only few tiny details.

**Comment by hensldm** [comment]
Hello, this PR hasn't been updated in a while. Do you have plans to update this PR? If we haven't heard back by September 1st then we will take it over / close the PR.

**Comment by hensldm** [comment]
Closing due to inactivity.
